### DISCUSSION TITLE
How to convert `TypedDict` to a `dataclass_transform` behavior?

### DISCUSSION THREAD

**Discussion by Kludex** [python/typing#1457] [Q&A]
Hi there 👋 

I'm one of the maintainers of Starlette, and I need help.

In Starlette ([ASGI]), we have _Lifespan events_ (startup and shutdown). There, you can store data that will be used on your endpoints (see line `yield {"potato": "potato"}` below for how to store).

And... There are two ways to retrieve this `potato` in the endpoint:
1. `request.scope["state"]["potato"]`, since `state` is a dictionary.
2. `request.state.potato`, since `Request.state` actually wraps the above dictionary, see [Request]/[State] for more details.

The first is an [ASGI] concept, and the second is specific to Starlette. Deprecating the second is not an option, but we can tweak the implementation as much as wanted if it doesn't break our users' applications.

Anyway... Is there a way to type hint the [Request]/[State] in a way the following `reveal_type`s shows the type of `potato`?

```py
from contextlib import asynccontextmanager
from typing import AsyncGenerator, TypedDict

from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import Response
from starlette.routing import Route


class LifespanState(TypedDict):
    potato: str


@asynccontextmanager
async def lifespan(app: Starlette) -> AsyncGenerator[LifespanState, None]:
    yield {"potato": "potato"}


async def homepage(request: Request) -> Response:
    # I want `request.state.potato` to be `str`.
    reveal_type(request.state.potato)

    # I also want `request.scope["state"]["potato"]` to be `str`.
    reveal_type(request.scope["state"]["potato"])

    return Response(content="Hello, world!")


app = Starlette(routes=[Route("/", endpoint=homepage)])
``` 

> [!Note]
> Title may be misleading - change it if you want. 👀 

[ASGI]: https://asgi.readthedocs.io/en/latest/
[Request]: https://github.com/encode/starlette/blob/master/starlette/requests.py
[State]: https://github.com/encode/starlette/blob/cb6d284877cd359a21a0f31ebf8103a7b285a848/starlette/datastructures.py#L682-L707

**Comment by developoohs** [comment]


**Reply by Kludex** [reply]
No. It doesn't solve the issue.

**Comment by sobolevn** [comment]
You can make `request.scope["state"]` generic, but you cannot convert `TypedDict` to a dataclass.

Here's how `.scope['state']` would work:

```python
_ScopeStateT = TypeVar('_ScopeStateT')

class Scope(TypedDict, Generic[_ScopeStateT]):
     state: _ScopeStateT

class Request(Generic[_ScopeStateT]):
     scope: Scope[_ScopeStateT]
```

And then:

```python
async def homepage(request: Request[LifespanState]) -> Response:
    reveal_type(request.scope["state"]["potato"])  # N: Revealed type is "builtins.str"
```

**Reply by Kludex** [reply]
Yeah, that's fine for the first case. I should have been more explicit that I'm looking to satisfy the second case.

> [...] but you cannot convert `TypedDict` to a dataclass.

Yeah... I know, but I don't know why. Would you mind sharing the reason? Also, I'm more interested in finding the less disruptive solution here that will make `request.state` type safe.

(Thanks @sobolevn 🙏)

**Reply by Daverball** [reply]
This just wasn't part of PEP-681's scope. So it's not possible right now without adding a second `TypeVar` for the other way to access the data. So the best you can do right now is make `request.state` an `Any` or a `Protocol` with a `__getattr__` which returns `Any`.

The following proposal might make something along the lines of what you want possible: https://github.com/python/typing/discussions/1412

**Reply by sobolevn** [reply]
You motivated me to add this support to https://github.com/wemake-services/mypy-extras/
You can watch it for the next releases.

**Reply by Kludex** [reply]
> This just wasn't part of PEP-681's scope. So it's not possible right now without adding a second `TypeVar` for the other way to access the data. So the best you can do right now is make `request.state` an `Any` or a `Protocol` with a `__getattr__` which returns `Any`.

This doesn't look like an acceptable API. 

> The following proposal might make something along the lines of what you want possible: #1412

This does look like something that will solve my issue! 🤩 Thanks!

**Reply by Kludex** [reply]
> You motivated me to add this support to [wemake-services/mypy-extras](https://github.com/wemake-services/mypy-extras/) You can watch it for the next releases.

Watching! Thanks! ❤️

**Comment by Kludex** [comment]
I really want to improve the experience of FastAPI/Starlette users, and having something like this would be great.
